crates/codegen/parser/runtime/src/query_engine.rs

```diff
@@ -0,0 +1,439 @@
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use super::cst;
+use super::cursor::Cursor;
+use super::query_model::{
+    AlternativesMatch, BindingMatch, Kind, NodeId, NodeMatch, OneOrMoreMatch, OptionalMatch, Query,
+    SequenceMatch,
+};
+
+impl Cursor {
+    pub fn query(self, queries: Vec<Query>) -> QueryIterator {
+        QueryIterator::new(self, queries)
+    }
+
+    fn consume(&mut self) -> bool {
+        if self.is_completed() {
+            false
+        } else {
+            if !self.go_to_next_sibling() {
+                self.complete();
+            }
+            true
+        }
+    }
+
+    fn matches_query_node_id(&self, node_id: &NodeId) -> bool {
+        match &self.current.node {
+            cst::Node::Rule(rule) => match node_id {
+                NodeId::Anonymous => true,
+                NodeId::Kind { kind } => Kind::Rule(rule.kind) == *kind,
+                NodeId::String { .. } => false,
+                NodeId::Field { field } => Some(*field) == self.node_name(),
+                NodeId::FieldAndKind { field, kind } => {
+                    Some(*field) == self.node_name() && Kind::Rule(rule.kind) == *kind
+                }
+                NodeId::FieldAndString { .. } => false,
+            },
+
+            cst::Node::Token(token) => match node_id {
+                NodeId::Anonymous => true,
+                NodeId::Kind { kind } => Kind::Token(token.kind) == *kind,
+                NodeId::String { string } => token.text == *string,
+                NodeId::Field { field } => Some(*field) == self.node_name(),
+                NodeId::FieldAndKind { field, kind } => {
+                    Some(*field) == self.node_name() && Kind::Token(token.kind) == *kind
+                }
+                NodeId::FieldAndString { field, string } => {
+                    Some(*field) == self.node_name() && token.text == *string
+                }
+            },
+        }
+    }
+}
+
+impl Query {
+    fn create_qecref(&self, cursor: Cursor) -> QECRef {
```

Comment:
TOL, I'm a bit confused by using `QecRef` (`QueryCombinator`?) vs `BindingCombinator` vs `QueryEngineCombinator` vs `QueryIterator` .. but maybe I need to spend some time using these types to have a better understanding ..

Reply:
QueryEngineCombinatorRef

Reply:
Open to suggestions. This is about as simple as I could get, and it took months to get here ... but always happy to improve naming. Maybe I'm being too concise, and the names should be more explicit?

Reply:
I changed to `Combinator` and `CombinatorRef`, because the types are local to the module and the names can therefore be scoped.

Reply:
Sounds great! thanks1